probe/main.go

```diff
@@ -55,129 +54,152 @@ func main() {
-			log.Print(err)
-		}()
 	}
 
-	if *spyProcs && os.Getegid() != 0 {
-		log.Printf("warning: process reporting enabled, but that requires root to find everything")
+	if len(flag.Args()) > 0 {
+		targets = flag.Args()
 	}
+	log.Printf("publishing to: %s", strings.Join(targets, ", "))
 
-	publisherFactory := func(target string) (xfer.Publisher, error) {
+	factory := func(target string) (xfer.Publisher, error) {
 		publisher, err := xfer.NewHTTPPublisher(target, *token, probeID)
 		if err != nil {
 			return nil, err
 		}
 		return xfer.NewBackgroundPublisher(publisher), nil
 	}
-	publishers := xfer.NewMultiPublisher(publisherFactory)
+
+	publishers := xfer.NewMultiPublisher(factory)
 	defer publishers.Stop()
+
 	resolver := newStaticResolver(targets, publishers.Add)
 	defer resolver.Stop()
 
-	addrs, err := net.InterfaceAddrs()
-	if err != nil {
-		log.Fatal(err)
-	}
-	localNets := report.Networks{}
-	for _, addr := range addrs {
-		// Not all addrs are IPNets.
-		if ipNet, ok := addr.(*net.IPNet); ok {
-			localNets = append(localNets, ipNet)
-		}
-	}
+	endpointReporter := endpoint.NewReporter(hostID, hostName, *spyProcs, *useConntrack)
+	defer endpointReporter.Stop()
+
+	processCache := process.NewCachingWalker(process.NewWalker(*procRoot))
 
 	var (
-		endpointReporter = endpoint.NewReporter(hostID, hostName, *spyProcs, *useConntrack)
-		processCache     = process.NewCachingWalker(process.NewWalker(*procRoot))
-		tickers          = []Ticker{processCache}
-		reporters        = []Reporter{
-			endpointReporter,
-			host.NewReporter(hostID, hostName, localNets),
-			process.NewReporter(processCache, hostID),
-		}
-		taggers = []Tagger{newTopologyTagger(), host.NewTagger(hostID)}
+		tickers   = []Ticker{processCache}
+		reporters = []Reporter{endpointReporter, host.NewReporter(hostID, hostName, localNets), process.NewReporter(processCache, hostID)}
+		taggers   = []Tagger{newTopologyTagger(), host.NewTagger(hostID)}
 	)
-	defer endpointReporter.Stop()
 
-	if *dockerEnabled {
+	dockerTagger, dockerReporter, dockerRegistry := func() (*docker.Tagger, *docker.Reporter, docker.Registry) {
+		if !*dockerEnabled {
+			return nil, nil, nil
+		}
 		if err := report.AddLocalBridge(*dockerBridge); err != nil {
-			log.Fatalf("failed to get docker bridge address: %v", err)
+			log.Printf("Docker: problem with bridge %s: %v", *dockerBridge, err)
+			return nil, nil, nil
 		}
-
-		dockerRegistry, err := docker.NewRegistry(*dockerInterval)
+		registry, err := docker.NewRegistry(*dockerInterval)
 		if err != nil {
-			log.Fatalf("failed to start docker registry: %v", err)
+			log.Printf("Docker: failed to start registry: %v", err)
+			return nil, nil, nil
 		}
+		return docker.NewTagger(registry, processCache), docker.NewReporter(registry, hostID), registry
+	}()
+	if dockerTagger != nil {
+		taggers = append(taggers, dockerTagger)
+	}
+	if dockerReporter != nil {
+		reporters = append(reporters, dockerReporter)
+	}
+	if dockerRegistry != nil {
```

Comment:
Can you hoist this stuff into the function?  the function has access to the taggers, reporters etc.  Also, why is the function even necessary?  Why not fatalf?